abi/codecForCompositeTypes_option.go

```diff
@@ -0,0 +1,65 @@
+package abi
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+)
+
+func (c *codec) encodeNestedOption(writer io.Writer, value OptionValue) error {
+	if value.Value == nil {
+		_, err := writer.Write([]byte{0})
+		return err
+	}
+
+	_, err := writer.Write([]byte{1})
+	if err != nil {
+		return err
+	}
+
+	return c.doEncodeNested(writer, value.Value)
+}
+
+func (c *codec) encodeTopLevelOption(writer io.Writer, value OptionValue) error {
+	if value.Value == nil {
+		return nil
+	}
+
+	_, err := writer.Write([]byte{1})
+	if err != nil {
+		return err
+	}
+
+	return c.doEncodeNested(writer, value.Value)
+}
+
+func (c *codec) decodeNestedOption(reader io.Reader, value *OptionValue) error {
+	bytes, err := readBytesExactly(reader, 1)
+	if err != nil {
+		return err
+	}
+
+	if bytes[0] == 0 {
+		value.Value = nil
+		return nil
+	}
+
```

Comment:
If you do it like this, you need to check bytes[0] == 1 otherwise return error.
Read below comment with my suggestion

Reply:
Yes, it is safer to add the check

Reply:
Indeed, added.